src/serac/numerics/assembled_sparse_matrix.cpp

```diff
@@ -0,0 +1,263 @@
+          // we haven't seen this before
+          trial_vdofs[nnz_row] = trial_j_vdof;
+          nnz_row++;
+        }
+      }
+    }
+
+    // add entries to I
+    for (int vi = 0; vi < test_vdim; vi++) {
+      const auto nnz_index_v = test_fes_.DofToVDof(test_vdof, vi);
+      const auto nnz_index   = nnz_index_v >= 0 ? nnz_index_v : -1 - nnz_index_v;
+      I[nnz_index]           = nnz_row * trial_vdim;
+    }
+  }
+
+  // Perform inclusive scan on all entries
+  int nnz = 0;
+  for (int i = 0; i < test_ndofs * trial_vdim; i++) {
+    int temp = I[i];
+    I[i]     = nnz;
+    nnz += temp;
+  }
+  I[test_ndofs * trial_vdim] = nnz;
+
+  return nnz;
+}
+
+void AssembledSparseMatrix::FillJ()
+{
+  auto& test_offsets    = detail::ElementRestriction_offsets(test_restriction_);
+  auto& test_indices    = detail::ElementRestriction_indices(test_restriction_);
+  auto& trial_gatherMap = detail::ElementRestriction_gatherMap(trial_restriction_);
+
+  const int test_elem_dof  = test_fes_.GetFE(0)->GetDof();
+  const int trial_elem_dof = trial_fes_.GetFE(0)->GetDof();
+  const int test_vdim      = test_fes_.GetVDim();
+  const int trial_vdim     = trial_fes_.GetVDim();
+
+  const int ne = trial_fes_.GetNE();
+  ea_map_.SetSize(test_elem_dof * test_vdim * trial_elem_dof * trial_vdim * ne);
+  auto map_ea = mfem::Reshape(ea_map_.ReadWrite(), test_elem_dof * test_vdim, trial_elem_dof * trial_vdim, ne);
+
+  // initialize J
+  for (int j = 0; j < this->J.Capacity(); j++) {
+    this->J[j] = -1;
+  }
+
+  for (int test_vdof = 0; test_vdof < test_fes_.GetNDofs(); test_vdof++) {
+    // Look through each element corresponding to a test_vdof
+    const int test_row_offset = test_offsets[test_vdof];
+    const int nrow_elems      = test_offsets[test_vdof + 1] - test_row_offset;
+
+    // here we assume all the components have the same number of columns
+    const int        nnz_row = I[test_fes_.DofToVDof(test_vdof, 0) + 1] - I[test_fes_.DofToVDof(test_vdof, 0)];
+    mfem::Array<int> trial_vdofs(nnz_row);
+    trial_vdofs      = -1;
+    int j_vdof_index = 0;
+
+    // Build temporary array for assembled J
+    for (int e_index = 0; e_index < nrow_elems; e_index++) {
+      // test_indices can be negative in the case of Hcurl
+      const int test_index_v = test_indices[test_row_offset + e_index];
+      const int test_index   = test_index_v >= 0 ? test_index_v : -test_index_v - 1;
+      const int e            = test_index / test_elem_dof;
+      const int test_i_elem  = test_index % test_elem_dof;
+
+      // find corresponding trial_vdofs
+      mfem::Array<int> trial_elem_vdofs(trial_elem_dof);
+      for (int j_elem = 0; j_elem < trial_elem_dof; j_elem++) {
+        // could be negative.. but trial_elem_vdofs is a temporary array
+        const auto trial_j_vdof_v = trial_gatherMap[trial_elem_dof * e + j_elem];
+        const auto trial_j_vdof   = trial_j_vdof_v >= 0 ? trial_j_vdof_v : -1 - trial_j_vdof_v;
```

Comment:
This same ternary construction shows up all over mfem too. I wish this weird sign encoding was encapsulated in a type rather than just being copy and pasted everywhere (e.g. https://godbolt.org/z/xGhnbMefo) 

That is probably out of our control though, since we're using mfem's data types.

Reply:
I think we can add it in our side. That's a good idea to tidy up the code.

Reply:
I've added some temporary sign-related templated methods just to help document the code a little more clearly.